Several smaller improvements for EDS HTML Full Text Rendering

cwolfebsco opened the pull request:
EDS offers HTML full text as part of the retrieve response for non-guest users. EDS RecordDriver utilizes a **_toHTML_** function which turns the encountered markup response into something that can render in HTML. This approach has at least three items that could be improved:

(1) The value of the `<anid>` tag will show -> this PR gives this a class (eds_html_anid) allowing for easy hiding via CSS
(2) Tables will show as raw HTML. These tables are enclosed in an <ephtml> tag
(3) Mathematical formulas will show as raw HTLM. These math formulas are also enclosed in an <ephtml> tag

The approach chosen here is flawed, but resolves these display issues. 

<img width="514" alt="Screenshot 2025-04-24 at 09 19 53" src="https://github.com/user-attachments/assets/410203c4-d416-4af9-a3b5-2e6f79bb7275" />

<img width="856" alt="Screenshot 2025-04-24 at 09 20 05" src="https://github.com/user-attachments/assets/aa5c16c7-2d8f-41a9-b4c5-8ff6b0171685" />

cwolfebsco committed:
Empower sites to hide the ANID string

EDS API HTML FT will contain a string on top that contains various details such as Accession Number & dates. This string might be something that some sides would rather hide.

The toHTML function converts the <anid> tag into a <p> tag. Adding the "eds_html_anid" class, will allow sites to hide this string via CSS, if they chose to do so

cwolfebsco committed:
render tables as tables rather than raw html

Tables would show as raw html, this attempts to split the string at ephtml and if the text starts with table it'll be wrapped in a div with class eds_html_table and the data will be run through html_entity_decode

We only bother if the string contains ephtml

cwolfebsco committed:
Ensuring mathematical annotations show

this will help mathematical annotations show, if they are created in MathML and if your browser supports it

cwolfebsco committed:
correcting comment

cwolfebsco committed:
clarity for variable name

cwolfebsco committed:
math formulars often used inline

cwolfebsco committed:
code formatting changes

cwolfebsco committed:
one more set of whitespace changes :)

cwolfebsco committed:
fix variable name

cwolfebsco commented:
@demiankatz I would appreciate your thoughts on this one and would be happy to discuss in detail. I left it as a draft pr, to give us an opportunity to consider the path to take first.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
```diff
@@ -660,8 +660,27 @@ protected function toHTML($data, $group = null)
+                $splitParts = preg_split($pattern, $data);
+                $data = '';
+                foreach ($splitParts as $part) {
+                    if (strpos($part, 'table') > -1 && strpos($part, 'table') < 6) {
```
What is the purpose of the "magic number" 6 here?

cwolfebsco replied:
Great question and a place where I should improve. The data will look like this:

`<ephtml> &lt;table&gt;&lt;tbody valign="top"&gt; ...`

After the preg_split it should be:

` &lt;table&gt;&lt;tbody valign="top"&gt; ...`

I probably should simply look for `&lt;table` and then just accept it if it is > -1. For math I could look for `&lt;math`.

I'll look into adding a test. you have been most helpful, as always!

demiankatz replied:
Yes, I think that your proposed change sounds reasonable and should make the code easier to understand. Thanks!

cwolfebsco replied:
@demiankatz I did push this change, sorry it took me so long

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco!

I have one question below.

In addition to that, I think it might be valuable to add one or more test cases to cover this new functionality in the [EDS RecordDriver unit test](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/tests/unit-tests/src/VuFindTest/RecordDriver/EDSTest.php). It would require adding a new [eds fixture](https://github.com/vufind-org/vufind/tree/dev/module/VuFind/tests/fixtures/eds) demonstrating the functionality; I assume it would not be too hard to create a fictionalized example without needing to share anything copyrighted.

What do you think? Let me know if I can do more to help with any of this!

cwolfebsco committed:
EDS Retrieve Response, single record, HTML + PDF FT

This contains `ephtml` tags for both tables and mathML

cwolfebsco committed:
correct JSON sample

cwolfebsco committed:
add basic test for the EPHTML parsing

cwolfebsco committed:
slight update to test

cwolfebsco committed:
fix whitespace issue, but long line warnings remain

cwolfebsco committed:
move long string to fixture

cwolfebsco committed:
enhance stropos to check for <table

rather than simply look for "table" and/or "math" we now look for &lt;table and/or &lt;math indicating the start of either a table or math formula

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
```diff
@@ -660,8 +660,27 @@ protected function toHTML($data, $group = null)
+
+            // ensure tables & mathML are not shown as raw html
+            if (strpos($data, 'ephtml') > -1) {
+                $pattern = "/<\/?ephtml>/";
```
One point of confusion -- you appear to be splitting using `<ephtml>` or `</ephtml>`, but doesn't the data actually contain `&lt;ephtml&gt;` or `&lt;/ephtml&gt;` at this point? I realize the test is passing, so I must be overlooking/misunderstanding something -- but I'm not sure I understand why we're using unencoded HTML in the regular expression but encoded HTML in the strpos statements below. If nothing else, maybe we need a comment to clarify this so I don't get confused by it again in the future. :-)

cwolfebsco replied:
I'll need to create a proper comment then.
Basically the data looks like this at this point:

`<epthml>&lt;math&gt;LOTS OF COMPLEX STUFF&lt;/math&gt;</ephtml>`

So we are splitting at ephtml, then URL decode, then handle from there...

demiankatz replied:
Ahh, I understand -- I got confused because looking at the JSON fixture, I saw that everything was encoded... but now I see that some things are DOUBLE encoded. That's where I went wrong. :-)

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco -- one point is confusing me; it may be an issue with my brain rather than your code, but could you take a look at my comment below and clarify for me? ;-)

cwolfebsco committed:
Enhance comments for clarity and detail

previous comment was insufficient to explain the purpose of this section of code, I hope this adds more clarity

demiankatz reviewed (approved):
Thanks, @cwolfebsco!
